SimTKcommon/src/Pathname.cpp

```diff
@@ -238,6 +240,161 @@ void Pathname::deconstructPathname( const string&   name,
     }
 }
 
+void Pathname::deconstructPathnameUsingSpecifiedWorkingDirectory(const std::string& swd,
+                                                                 const std::string& path,
+                                                                 std::string& directory,
+                                                                 std::string& fileName,
+                                                                 std::string& extension)
+{
+    directory.erase(); fileName.erase(); extension.erase();
+    string pathdrive, swddrive, finaldrive;
+    String processed;
+
+    // Remove all the white space and make all the slashes be forward ones.
+    // (For Windows they'll be changed to backslashes later.)
+    String pathcleaned = String::trimWhiteSpace(path).replaceAllChar('\\', '/');
+    if (pathcleaned.empty())
+        return; // path consisted only of white space
+    removeDriveInPlace(pathcleaned, pathdrive);
+
+    String swdcleaned = String::trimWhiteSpace(swd).replaceAllChar('\\', '/');
+    if (!swdcleaned.empty() && swdcleaned[swdcleaned.size() - 1] != '/')
+        swdcleaned += '/';
+    removeDriveInPlace(swdcleaned, swddrive);
+
+    // If swd is empty (or only white space) then use deconstructPathname().
+    if (swdcleaned.empty() && swddrive.empty()) {
+        bool dontApplySearchPath;
+        deconstructPathname(path, dontApplySearchPath, directory, fileName, extension);
+        if (!dontApplySearchPath)
+            directory = getCurrentWorkingDirectory() + directory;
+
+    // swd was not empty. If necessary, preprocess the swd. Then concatenate swd and path.
+    // Finally, resolve drives.
+    }
+    else {
+        /* PREPROCESSING THE SWD. */
+        // Preprocess the swd if it leads with a "./"
+        if (swdcleaned.substr(0, 2) == "./") {
+            swdcleaned.replace(0, 2, getCurrentWorkingDirectory(swddrive));
+            removeDriveInPlace(swdcleaned, swddrive);
+            swdcleaned.insert(0, "/"); // ensure swd starts with "/" to denote full path
+        }
+        // Also preprocess the swd if it leads with "/". Just grab current drive letter.
+        else if (swdcleaned.substr(0, 1) == "/" && swddrive.empty()) {
+            swddrive = getCurrentDriveLetter();
+        }
+        // Also preprocess if swd starts with something of the form "C:folder/". Resolve
+        // by finding the current directory of this drive.
+        else if (!swddrive.empty()) {
+            swdcleaned.insert(0, getCurrentWorkingDirectory(swddrive));
+            removeDriveInPlace(swdcleaned, swddrive);
+            swdcleaned.insert(0, "/"); // ensure swd starts with "/" to denote full path
```

Comment:
getCurrentWorkingDirectory() does return an absolute path, but removeDriveInPlace() does a str.erase(0,2), so it removes the leading "/". I considered just erasing the first two by a direct call instead of using the helper function, but that wouldn't work correctly on not-Windows platforms.